app/scripts/controllers/permissions/methodMiddleware.js

```diff
@@ -81,10 +84,33 @@ export default function createMethodMiddleware ({
         res.result = true
         return
 
+      // register return handler to send accountsChanged notification
+      case 'wallet_requestPermissions':
+
+        if ('eth_accounts' in req.params?.[0]) {
+
+          responseHandler = async () => {
+
+            if (Array.isArray(res.result)) {
+              for (const permission of res.result) {
+                if (permission.parentCapability === 'eth_accounts') {
+                  notifyAccountsChanged(await getAccounts())
+                }
+              }
+            }
+          }
+        }
+        break
+
       default:
         break
     }
 
-    next()
+    // when this promise resolves, the response is on its way back
+    await next()
+
+    if (responseHandler) {
+      responseHandler()
+    }
```

Comment:
This how to do response/return handlers for `createAsyncMiddleware` middlewares. Here, `next` does not take a return handler callback like in synchronous middlewares.